-lオプション付きのlsコマンドを作成しました。

sht3ng opened the pull request.

sht3ng committed:
ls.rbとls_test.rbのコピー（ls4.rbとls4_test.rb）を作成しました。

sht3ng committed:
ls4_test.rbを削除しました。

sht3ng committed:
lオプションつきのlsコマンドが機能する状態にしました。

sht3ng committed:
rubocop 関数の複雑性が高い問題と一行が長すぎる問題を修正しました。

sht3ng committed:
rubocop コレクタブルな表記上の軽微な問題を修正しました。

sht3ng committed:
rubocop 関数の複雑性の問題を解決しました。

sht3ng committed:
=の位置を揃えたり、if-elseをcase-whenに変更したり軽微な修正を行いました。

yuuu commented on 05.ls/ls4.rb:
```diff
@@ -0,0 +1,222 @@
+
+def read_file_stats(path, files)
+  file_stats = []
+  files.each do |file|
```
ここは `files` という `Array` から `file_stats` という別の `Array` を作っている処理なので、`Array#map` を使ったほうが良いですね。

yuuu commented on 05.ls/ls4.rb:
```diff
@@ -0,0 +1,222 @@
+
+def translate_octal_permissions_into_symbols(octal_permissions, octal_special_permissions)
+  symbolic_permissions = []
+  octal_permissions.chars.each do |octal_permission|
```
ここも `Array#map` を使ったほうが良いですね 👍

yuuu replied:
こちらも同様です。
https://github.com/sht3ng/ruby-practices/pull/7/files#diff-df06c71a195d7f95323c229339a664594ff4ad7b99e66a11d01168fccd9f0032R148

yuuu commented on 05.ls/ls4.rb:
```diff
@@ -0,0 +1,222 @@
+  symbolic_permissions
+end
+
+def overwrite_symbols_with_suid(symbolic_permissions, suid_permission)
```
以下3つのメソッドは配列の参照位置と最終的な文字列が異なるだけに見えます。
1つのメソッドにまとめることは難しいでしょうか？

- overwrite_symbols_with_suid
- overwrite_symbols_with_sgid
- overwrite_symbols_with_stickybit

ちなみに文字の大文字・小文字を切り替えるメソッドとしてこういったものもあります。
https://docs.ruby-lang.org/ja/latest/method/String/i/upcase.html

yuuu commented on 05.ls/ls4.rb:
```diff
@@ -0,0 +1,222 @@
+
+def generate_timestamp(time)
+  timestamp = []
+  timestamp << time.strftime('%_m')
```
少なくともこの前後の行は、まとめて

```ruby
timestamp = [time.strftime('%_m%_d')]
```

とすることができますね。

yuuu commented on 05.ls/ls4.rb:
```diff
@@ -0,0 +1,222 @@
+end
+
+def generate_file_name(path, file)
+  return "#{file} -> #{File.readlink("#{path}/#{file}")}" if File.symlink?("#{path}/#{file}")
```
シンボリックリンクの表示に対応していますね 👍

sht3ng committed:
Array#eachを使用していた箇所をArray#mapに変更しました。

sht3ng committed:
timestampの月と日を一つの行で作成するように変更しました。

sht3ng committed:
特殊なアクセス権の上書き関数を一つにまとめました。

sht3ng committed:
Array#mapのレシーバの配列名とブロック変数名が被っていた問題を修正しました。

sht3ng commented:
@yuuu さん
ご確認くださってありがとうございます！☺️
頂いたコメントを参考に修正いたしましたので、
ご確認の程よろしくお願いいたします🙇‍♂️

yuuu commented on 05.ls/ls4.rb:
```diff
@@ -127,83 +124,49 @@ def translate_octal_type_into_symbol(octal_type)
-      'T'
-    end
+  list_of_special_permission_symbols = %w[s s t]
+  binary_special_permission.chars.each_with_index do |permission, index|
```
さらにリファクタリングができそうです 👍 

- ここは `symbolic_permission` を書き換えるのではなく `Array#map` を使うことで新しい結果のArrayを作る方針にするとコードがスッキリすると思います
- `permission == '1'` の記述が重複しているので、先に `if permission == '0'` をして条件を満たすなら文字をそのまま返すようにすると良さそうです。
    - [ここ](https://github.com/sht3ng/ruby-practices/pull/7/files/06b63722c1084e5d5ec54b84926848d953f1e614..0bac1c40ce32bd0f250582703cf16c043aee3d94#diff-df06c71a195d7f95323c229339a664594ff4ad7b99e66a11d01168fccd9f0032R152) のアーリーリターンは不要になりますね。
    - 一旦 `list_of_special_permission_symbols[index]` をセットして、 `symbolic_permission[index][2] == '-'` を満たすなら `upcase` する、という書き方だとスッキリしそうですね。

----

まとめるとこういうイメージです。「これだと上手くいかない」などありましたら教えて下さい🙏

```ruby
binary_special_permission.chars.map.with_index do |permission, index|
  if permission == '0'
    # そのままの値を返す
  else
    symbol = list_of_special_permission_symbols[index]
    symbolic_permission[index][2] == '-' ? symbol.upcase : synbol
  end
end
```

sht3ng committed:
特殊なアクセス権の関数内の処理をArray#eachからArray#mapへ変更しました。

sht3ng commented:
@yuuu さん
再確認くださりありがとうございます☺️
頂いたアドバイスをもとにリファクタリングを行いました！
コードがさらにすっきりしてわかりやすくなったように思います✨
お手数をおかけして恐縮ですが、ご確認よろしくお願いいたします🙇‍♂️

sht3ng committed:
インデントの調整を行いました。

sht3ng committed:
ロング形式とデフォルト形式それぞれの表示コマンドの関数名を変更しました。

ls_with_option_l -> ls_in_long_format
ls_without_options -> ls_in_default_format
